### PR TITLE
shows mr-x location on the map on reveal moves

### PR DESCRIPTION
**Pull request by azharkhan**
- sets Map prop to show Mr.X location

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -7,10 +7,6 @@
     <link rel="stylesheet" href="./src/styles.css" />
   </head>
   <body>
-    <div id="app">
-
-    </div>
-    <script src="./src/app.js"></script>
```
Comment (azharkhan) [inline index.html:13]: these are extraneous

File: src/components/Map.vue [hunk 2]
```diff
@@ -49,6 +49,8 @@ export default {
   props: {
     detectives: Array,
     currentPlayer: Object || null,
+    isMrXVisible: Boolean,
+    mrXLocation: Number || Boolean,
```
Comment (azharkhan) [inline src/components/Map.vue:53]: it can be false, when Mr.X is not visible, so we need to account for the extra type

File: index.html (hunk 1 above)
Comment (neaprabhu) [reply index.html:13]: Makes sense

File: src/components/Map.vue (hunk 2 above)
Comment (neaprabhu) [reply src/components/Map.vue:53]: instead of tagging the boolean to mrXLocation, can isMrXVisible itself be used?

File: src/components/Map.vue [hunk 3]
```diff
@@ -264,9 +265,7 @@ export default {
         this.clearAvailableRoutes();
 
         this.detectives.map(detective => {
-          const point = document.querySelector(
-            `[data-point="${detective.currentLocation.toString()}"]`
-          );
+          const point = this.$refs[detective.currentLocation][0];
```
Comment (neaprabhu) [inline src/components/Map.vue:268]: Bro can you explain that this line of code is doing? From what I know, $refs are used to access elements of the DOM that have a ref attribute in their tags right? eg: <div ref="mydiv">

$ref.mydiv would access the above div

Comment (neaprabhu) [reply src/components/Map.vue:268]: Since detective is passed in to the component as a prop, is there a reason why you wouldnt access the current location directly? as detective.currentLocation[0] ?

File: src/components/Map.vue [hunk 4]
```diff
@@ -276,15 +275,39 @@ export default {
...
+    isMrXVisible: {
+      handler() {
+        const svg = document.getElementById("board");
+        if (this.isMrXVisible) {
+          // check if user has access to Mr X
+          const isMrXTurn =
+            this.currentPlayer && this.currentPlayer.role === "mr-x";
+
+          if (!isMrXTurn && this.isMrXVisible && this.mrXLocation) {
+            const mrXLocationStation = this.$refs[this.mrXLocation][0];
```
Comment (neaprabhu) [inline src/components/Map.vue:287]: Bro can you explain this line in detail?

File: src/components/Map.vue [hunk 5]
```diff
@@ -276,15 +275,39 @@ export default {
...
+            const attributes = {
+              cx: mrXLocationStation.getAttributeNS(null, "cx"),
+              cy: mrXLocationStation.getAttributeNS(null, "cy"),
+              className: `mr-x-location`,
+            };
+            this.addCircleToBoard(attributes);
+          }
+        } else {
```
Comment (neaprabhu) [inline src/components/Map.vue:295]: So this portion assumes that is a previous marker on the map for Mr X and you are now removing it. So for example on move # 4, you would enter the else portion of the if - else conditional, and remove the X marker that was added in move # 3

File: src/components/Map.vue [hunk 6]
```diff
@@ -276,15 +275,39 @@ export default {
...
+        this.setCurrentPlayer();
       },
     },
     currentPlayer: {
       handler() {
-        this.checkIfMrXVisible();
+        this.setCurrentPlayer();
       },
     },
   },
```
Comment (neaprabhu) [inline src/components/Map.vue:313]: Bro tbh not understanding, what these lines are doing? What is the handler function?

File: src/views/Game.vue [hunk 7]
```diff
@@ -100,6 +106,19 @@ export default {
     result: function() {
       return this.game && this.game.result;
     },
+    shouldRevealMrX: function() {
```
Comment (neaprabhu) [inline src/views/Game.vue:109]: This is basically being used to NOT reveal mr X right? method name could be misleading, that if I have understood the logic correctly

File: src/components/Map.vue (hunk 2 above)
Comment (azharkhan) [reply src/components/Map.vue:53]: unfortunately that would involve passing the location of Mr. X to the Map at all times, which I didn't want to do (not that there's any issue with it)

this seemed cleaner to me, since Mr. X location is either known (Number) or unknown (false Boolean)

but I can be convinced otherwise.

File: src/components/Map.vue (hunk 3 above)
Comment (azharkhan) [reply src/components/Map.vue:268]: Since I am creating a `ref` for each station by it's station number in the Map.vue component - https://github.com/azharkhan/scotland-yard/blob/master/src/components/Map.vue#L17

I can basically do the same thing I did above with the querySelector, but use `refs` for it. 

However, a small gotcha with setting refs in a `v-for` loop is that the ref is set as an Array - which is why I need to access the first index, even though there's only one ref in the Array for each station.

See Also: https://vuejs.org/v2/api/#ref

> When used on elements/components with v-for, the registered reference will be an Array containing DOM nodes or component instances.

File: src/components/Map.vue (hunk 4 above)
Comment (azharkhan) [reply src/components/Map.vue:287]: same as above, I'm basically getting the exact circle DOM element (which is stored as a `ref`) see comment above:

https://github.com/azharkhan/scotland-yard/blob/master/src/components/Map.vue#L17

File: src/components/Map.vue (hunk 6 above)
Comment (azharkhan) [reply src/components/Map.vue:313]: `watch` is basically a Vue component helper that gets called whenever the watched property is changed. the handler is basically, the function that triggers when that happens. 

see: https://vuejs.org/v2/api/#watch

I can simplify this a lot by just removing the handler function since it's shorthand calls the function defined as well as seen in the doc above. 

but in plain english - basically when the `state` changes (user is logged in or not) as well as when the currentPlayer changes (at the end of each player's turn) we want to call the `setCurrentPlayer` function.

File: src/views/Game.vue (hunk 7 above)
Comment (azharkhan) [reply src/views/Game.vue:109]: I think the intention was for this to return a Boolean (which it does using the `some` Array helper) [which will basically return true / false]

so the `shouldRevealMrX` will be `false` for all cases except when the moveNumber is one of the revealMoves.

but if I can name this anything better to be more reflective of it's intent, let me know.

File: src/components/Map.vue (hunk 5 above)
Comment (azharkhan) [reply src/components/Map.vue:295]: exactly, just in case there was a marker, we want to remove it - this is particularly helpful if Mr. X decides to make a 2x move after his reveal move.

**Comment by azharkhan** [comment]
comments have been replied to, happy to jump on a chat separately to talk about any of the specific questions :)
